src/terms.jl

```diff
@@ -0,0 +1,524 @@
+* `contrasts::ContrastsMatrix`: A contrasts matrix that captures the unique 
+  values this variable takes on and how they are mapped onto numerical 
+  predictors.
+"""
+struct CategoricalTerm{C,T,N} <: AbstractTerm
+    sym::Symbol
+    contrasts::ContrastsMatrix{C,T}
+end
+Base.show(io::IO, t::CategoricalTerm{C,T,N}) where {C,T,N} =
+    print(io, "$(t.sym) ($(length(t.contrasts.levels)) levels): $C($N)")
+width(::CategoricalTerm{C,T,N}) where {C,T,N} = N
+
+# constructor that computes the width based on the contrasts matrix
+CategoricalTerm(sym::Symbol, contrasts::ContrastsMatrix{C,T}) where {C,T} =
+    CategoricalTerm{C,T,length(contrasts.termnames)}(sym, contrasts)
+
+"""
+    MatrixTerm{Ts} <: AbstractTerm
+
+A collection of terms that should be combined to produce a single numeric matrix.
+
+A matrix term is created by [`apply_schema`](@ref) from a tuple of terms using 
+[`extract_matrix_terms`](@ref), which pulls out all the terms that are matrix
+terms as determined by the trait function [`is_matrix_term`](@ref), which is 
+true by default for all `AbstractTerm`s.
+"""
+struct MatrixTerm{Ts<:TupleTerm} <: AbstractTerm
+    terms::Ts
+end
+# wrap single terms in a tuple
+MatrixTerm(t::AbstractTerm) = MatrixTerm((t, ))
+
+Base.show(io::IO, t::MatrixTerm) = show(io, t.terms)
+width(t::MatrixTerm) = sum(width(tt) for tt in t.terms)
+
+"""
+    extract_matrix_terms(ts::TupleTerm)
+    extract_matrix_terms(t::AbstractTerm) = extract_matrix_term((t, ))
+
+Depending on whether the component terms are matrix terms (meaning they have
+[`is_matrix_term(T) == true`](@ref is_matrix_term)), `extract_matrix_terms` will
+return
+
+1.  A single `MatrixTerm` (if all components are matrix terms)
+2.  A tuple of the components (if none of them are matrix terms)
+3.  A tuple of terms, with all matrix terms collected into a single `MatrixTerm`
+    in the first element of the tuple, and the remaining non-matrix terms passed
+    through unchanged.
+
+By default all terms are matrix terms (that is,
+`is_matrix_term(::Type{<:AbstractTerm}) = true`), the first case is by far the
+most common.  The others are provided only for convenience when dealing with
+specialized terms that can't be concatenated into a single model matrix, like
+random effects terms in
+[MixedModels.jl](https://github.com/dmbates/MixedModels.jl).
+
+"""
+function extract_matrix_terms(ts::TupleTerm)
+    ismat = collect(is_matrix_term.(ts))
+    if all(ismat)
+        MatrixTerm(ts)
+    elseif any(ismat)
+        matterms = ts[ismat]
+        (MatrixTerm(ts[ismat]), ts[.!ismat]...)
+    else
+        ts
+    end
+end
+extract_matrix_terms(t::T) where {T<:AbstractTerm} =
+    is_matrix_term(T) ? MatrixTerm((t, )) : t
+extract_matrix_terms(t::MatrixTerm) = t
+
+
+"""
+    is_matrix_term(::Type{<:AbstractTerm})
+
+Does this type of term get concatenated with other matrix terms into a single
+model matrix?  This controls the behavior of the `MatrixTerm` constructor.  If
```

Comment:
Doesn't this describe `extract_matrix_terms` rather than the `MatrixTerm` constructor?

BTW, maybe `matrix_terms` would be enough (since "extract" doesn't add much information)?

Reply:
to me, `matrix_terms` sounds like it's like selecting or filtering for just the matrix terms.  I guess `collect_matrix_terms` is a better name for what this does...

Reply:
I don't think "extract" or "collect" help with the ambiguity. AFAICT the problem is that there's a possible confusion between `MatrixTerm` terms and `is_matrix_term` terms.

A possible way of limiting the confusion would be to adapt the terminology so that groups of terms (including `MatrixTerm` and `FormulaTerm`) have a different name from elementary terms.